# Cleaning Laptops Data

CSV file containing information about 1,300 laptop computers. The first five rows of the CSV file are shown below:

In [2]:
# import libraries
import pandas as pd

In [4]:
# data set
laptops = pd.read_csv('laptops.csv', encoding='latin-1')
laptops.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Manufacturer              1303 non-null   object
 1   Model Name                1303 non-null   object
 2   Category                  1303 non-null   object
 3   Screen Size               1303 non-null   object
 4   Screen                    1303 non-null   object
 5   CPU                       1303 non-null   object
 6   RAM                       1303 non-null   object
 7    Storage                  1303 non-null   object
 8   GPU                       1303 non-null   object
 9   Operating System          1303 non-null   object
 10  Operating System Version  1133 non-null   object
 11  Weight                    1303 non-null   object
 12  Price (Euros)             1303 non-null   object
dtypes: object(13)
memory usage: 132.5+ KB


We can see that every column is represented as the object type, indicating that they are represented by strings, not numbers. Also, one of the columns, Operating System Version, has null values.

The column labels have a variety of upper and lowercase letters, as well as spaces and parentheses, which will make them harder to work with and read. One noticeable issue is that the `" Storage"` column name has a space in front of it.

In [3]:
# old column
old_columns = laptops.columns
print('Old Columns: ', old_columns)

# remove any whitespace of each column
new_columns = []
for col in old_columns:
    new_columns.append(col.strip())

print('New Columns: ', new_columns)   

Old Columns:  Index(['Manufacturer', 'Model Name', 'Category', 'Screen Size', 'Screen',
       'CPU', 'RAM', ' Storage', 'GPU', 'Operating System',
       'Operating System Version', 'Weight', 'Price (Euros)'],
      dtype='object')
New Columns:  ['Manufacturer', 'Model Name', 'Category', 'Screen Size', 'Screen', 'CPU', 'RAM', 'Storage', 'GPU', 'Operating System', 'Operating System Version', 'Weight', 'Price (Euros)']


However, the column labels still have a variety of upper and lowercase letters, as well as parentheses, which will make them harder to work with and read. Let's finish cleaning our column labels by:

- Replacing spaces with underscores.
- Removing special characters.
- Making all labels lowercase.
- Shortening any long column names.

In [4]:
# function to clean columns labels
def clean_col(col):
    col = col.strip()
    col = col.replace('(','')
    col = col.replace(')','')
    col = col.lower()
    col = col.replace('Operating System', 'os')
    col = col.replace(' ','_')
    return col

In [5]:
# cleaning column labels
new_columns = []
for col in laptops.columns:
    new_label = clean_col(col)
    new_columns.append(new_label)
    
print('New Columns: ', new_columns)    

New Columns:  ['manufacturer', 'model_name', 'category', 'screen_size', 'screen', 'cpu', 'ram', 'storage', 'gpu', 'operating_system', 'operating_system_version', 'weight', 'price_euros']


In [6]:
laptops.columns = new_columns

We observed earlier that all 13 columns have the object dtype, meaning they're stored as strings. Let's look at the first few rows of some of our columns:

In [7]:
# first rows
print(laptops.iloc[:5,2:5])

    category screen_size                              screen
0  Ultrabook       13.3"  IPS Panel Retina Display 2560x1600
1  Ultrabook       13.3"                            1440x900
2   Notebook       15.6"                   Full HD 1920x1080
3  Ultrabook       15.4"  IPS Panel Retina Display 2880x1800
4  Ultrabook       13.3"  IPS Panel Retina Display 2560x1600


We want to explore data:

In [8]:
# explore the screen_size
print(laptops['screen_size'].dtype)
print(laptops['screen_size'].unique())

object
['13.3"' '15.6"' '15.4"' '14.0"' '12.0"' '11.6"' '17.3"' '10.1"' '13.5"'
 '12.5"' '13.0"' '18.4"' '13.9"' '12.3"' '17.0"' '15.0"' '14.1"' '11.3"']


All values in this column follow the same pattern, and there are no special cases. We will need to convert the column to a float type.

In [9]:
laptops['screen_size'] = laptops['screen_size'].str.replace('"','')
print(laptops['screen_size'].unique())

['13.3' '15.6' '15.4' '14.0' '12.0' '11.6' '17.3' '10.1' '13.5' '12.5'
 '13.0' '18.4' '13.9' '12.3' '17.0' '15.0' '14.1' '11.3']


In [10]:
laptops['screen_size'] = laptops['screen_size'].astype(float)
print(laptops['screen_size'].unique())

[13.3 15.6 15.4 14.  12.  11.6 17.3 10.1 13.5 12.5 13.  18.4 13.9 12.3
 17.  15.  14.1 11.3]


In [11]:
# explore ram column
print(laptops['ram'].dtype)
print(laptops['ram'].unique())

object
['8GB' '16GB' '4GB' '2GB' '12GB' '6GB' '32GB' '24GB' '64GB']


We identified a clear pattern in the `ram` column - all values are integers and include the character `GB` at the end of the string

In [12]:
laptops['ram'] = laptops['ram'].str.replace('GB','').astype(int)
print(laptops['ram'].unique())

[ 8 16  4  2 12  6 32 24 64]


We have converted some columns to numeric dtypes, and we need to rename the column because it is usefull if the non-digit values contain information that help us to understand the data

In [14]:
# rename screen_size
laptops.rename({'screen_size':'screen_size_inches'}, axis=1, inplace=True)

# rename ram
laptops.rename({'ram':'ram_gb'}, axis=1, inplace=True)

# describe
print(laptops['screen_size_inches'].describe())
print('\n')
print(laptops['ram_gb'].describe())

count    1303.000000
mean       15.017191
std         1.426304
min        10.100000
25%        14.000000
50%        15.600000
75%        15.600000
max        18.400000
Name: screen_size_inches, dtype: float64


count    1303.000000
mean        8.382195
std         5.084665
min         2.000000
25%         4.000000
50%         8.000000
75%         8.000000
max        64.000000
Name: ram_gb, dtype: float64


In [16]:
# data set
laptops.dtypes

manufacturer                 object
model_name                   object
category                     object
screen_size_inches          float64
screen                       object
cpu                          object
ram_gb                        int32
storage                      object
gpu                          object
operating_system             object
operating_system_version     object
weight                       object
price_euros                  object
dtype: object

We will analize the values in gpu and cpu columns

In [19]:
# first rows gpu
laptops['gpu'].head()

0    Intel Iris Plus Graphics 640
1          Intel HD Graphics 6000
2           Intel HD Graphics 620
3              AMD Radeon Pro 455
4    Intel Iris Plus Graphics 650
Name: gpu, dtype: object

The information in this columns seems to be a manufacturer(Intel, AMD) followed by a model name/number.

In [18]:
# gpu manufacturer
laptops['gpu_manufacturer'] = laptops['gpu'].str.split().str[0]
laptops['gpu_manufacturer'].head()

0    Intel
1    Intel
2    Intel
3      AMD
4    Intel
Name: gpu_manufacturer, dtype: object

In [23]:
# gpu manufacturers
laptops['gpu_manufacturer'].value_counts()

Intel     722
Nvidia    400
AMD       180
ARM         1
Name: gpu_manufacturer, dtype: int64

In [20]:
# first rows cpu
laptops['cpu'].head()

0          Intel Core i5 2.3GHz
1          Intel Core i5 1.8GHz
2    Intel Core i5 7200U 2.5GHz
3          Intel Core i7 2.7GHz
4          Intel Core i5 3.1GHz
Name: cpu, dtype: object

In [21]:
# cpu manufacturer
laptops['cpu_manufacturer'] = laptops['cpu'].str.split().str[0]
laptops['cpu_manufacturer'].head()

0    Intel
1    Intel
2    Intel
3    Intel
4    Intel
Name: cpu_manufacturer, dtype: object

In [25]:
# cpu manufacturers
laptops['cpu_manufacturer'].value_counts()

Intel      1240
AMD          62
Samsung       1
Name: cpu_manufacturer, dtype: int64

If your data has been scraped from a webpage or if there was manual data entry involved at some point, you may end up with inconsistent values. Let's look at an example from our os column:

In [32]:
# values operating system column
laptops['operating_system'].value_counts()

Windows      1125
No OS          66
Linux          62
Chrome OS      27
macOS          13
Mac OS          8
Android         2
Name: operating_system, dtype: int64

We can see that there are two variations of the Apple operating system — `macOS` — in our dataset: `Mac OS` and `macOS`

In [33]:
# corrected values in operating system column
mapping_dict = {
    'Android': 'Android',
    'Chrome OS': 'Chrome OS',
    'Linux': 'Linux',
    'Mac OS': 'macOS',
    'No OS': 'No OS',
    'Windows': 'Windows',
    'macOS': 'macOS'
}
# columns correction
laptops['operating_system'] = laptops['operating_system'].map(mapping_dict)

# corrected values operating system column
laptops['operating_system'].value_counts()

Windows      1125
No OS          66
Linux          62
Chrome OS      27
macOS          21
Android         2
Name: operating_system, dtype: int64

Dropping Missing Values

In [34]:
# cheking null values
laptops.isnull().sum()

manufacturer                  0
model_name                    0
category                      0
screen_size_inches            0
screen                        0
cpu                           0
ram_gb                        0
storage                       0
gpu                           0
operating_system              0
operating_system_version    170
weight                        0
price_euros                   0
gpu_manufacturer              0
cpu_manufacturer              0
dtype: int64

It's now clear that we have only one column with null values, `operating_system_version`, which has 170 missing values.

In [39]:
# no null rows
laptops_no_null_rows = laptops.dropna()
laptops_no_null_rows.shape

(1133, 15)

In [40]:
# no null columns
laptops_no_null_columns = laptops.dropna(axis=1)
laptops_no_null_columns.shape

(1303, 14)

Before dropping some values, we will explore `operating_system_version` columns to make a better decision. 

In [42]:
# value counts operating system version
laptops['operating_system_version'].value_counts(dropna=False)

10      1072
NaN      170
7         45
X          8
10 S       8
Name: operating_system_version, dtype: int64

We can see that the majority of values in the column are 10 and missing values are the next most common. Since `operating_system` is related to the `operating_system_version` columns, we will look at rows which the `operating_system_version` is missing

In [43]:
os_null_version = laptops.loc[laptops['operating_system_version'].isnull(), 'operating_system']

os_null_version.value_counts()

No OS        66
Linux        62
Chrome OS    27
macOS        13
Android       2
Name: operating_system, dtype: int64